src/Framework/PaymentGateways/Routes/GatewayRoute.php

```diff
@@ -86,4 +118,29 @@ private function isValidListener()
     {
         return isset($_GET['give-listener']) && $_GET['give-listener'] === 'give-gateway';
     }
+
+    /**
+     * Returns whether nonce is valid.
+     *
+     * @unreleased
+     *
+     * @since 2.18.0
+     *
+     * @return bool
+     */
+    private function hasValidNonce(GatewayRouteData $data)
+    {
+        // Few gateway route like offsite payment gateway redirect url will contain nonce
+        // Which we need to verify for security purpose.
+        if (null !== $data->nonce) {
+            return wp_verify_nonce(
+                $data->nonce,
+                (new GenerateGatewayRouteUrl())->getNonceActionName($data)
+            );
+        }
+
+        // We can return true for other gateway route urls like webhook, paypal ipn
+        // Because these routes will not contain nonce.
+        return true;
+    }
```

Comment:
Let's think through the best way for including and handling nonces in the route.   It's a little weird that this is completely optional and only validates when it's included.